lib/fs/walkfs.go

```diff
@@ -55,6 +87,21 @@ func (f *walkFilesystem) walk(path string, info FileInfo, walkFn WalkFunc) error
 		return nil
 	}
 
+	if ancestors.PushUnlessPresent(info) {
```

Comment:
Frankly I think this would be clearer as 

```
if ancestors.Contains(info) {
  return fmt.Errorf ...
}
ancestors.Push(info)
defer func() {
  ...
}()
```

that is, breaking out the double-action PushUnlessPresent into two operations and doing early return.

Or at the very least do just the early return to avoid the extra indent, but anyway.

Reply:
I'm still not clear on the circumstances under which the panic would happen, and it's a lot of machinery compared to just a `defer ancestors.Pop()` -- what's the rationale, really?